### PR TITLE
Fix bugprone-switch-missing-default-case

### PR DESCRIPTION
**Pull request by masterleinad**
Related to https://github.com/kokkos/kokkos/issues/7575.

### CODE REVIEW

**Review by nmm0 — APPROVED** [review]
Looks good to me!

**Comment by dalg24** [comment]
In gcc-smoketest
```
[ RUN      ] serial.atomic_operations_double
unreachable
Backtrace:
[0x558fdf69fae5] 
[0x558fdf67fc44] 
[0x558fdeb6cd45] 
[0x558fdeba9b07] 
[0x558fdeba975a] 
[0x558fdf6777bb] 
[0x558fdf65f3ea] 
[0x558fdf63a053] 
[0x558fdf63adb9] 
[0x558fdf63b5c4] 
[0x558fdf64bf10] 
[0x558fdf67a30b] 
[0x558fdf66180a] 
[0x558fdf64ba88] 
[0x558fdeb6b121] 
[0x558fdeb6b075] 
[0x7fd56c629d90] 
[0x7fd56c629e40] __libc_start_main
[0x558fdeb6af75]

**Review by JBludau — APPROVED** [review]
lgtm

File: containers/src/impl/Kokkos_Functional_impl.hpp [hunk 1]
```diff
@@ -89,6 +89,8 @@ uint32_t MurmurHash3_x86_32(const void* key, int len, uint32_t seed) {
       k1 = rotl32(k1, 15);
       k1 *= c2;
       h1 ^= k1;
+      break;
+    default: break;
```
Comment (dalg24) [inline containers/src/impl/Kokkos_Functional_impl.hpp:93]: I would just have suppressed the diagnostic for that one.

File: core/unit_test/TestAtomicOperations_double.hpp [hunk 2]
```diff
@@ -25,7 +25,7 @@ TEST(TEST_CATEGORY, atomic_operations_double) {
   const int start = -5;
   const int end   = 11;
   for (int i = start; i < end; ++i) {
-    for (int t = 0; t < 8; t++)
+    for (int t = 0; t < 7; t++)
```
Comment (dalg24) [inline core/unit_test/TestAtomicOperations_double.hpp:28]: OK.  Not loving how we do these but that is beyond the scope of this pr

**Review by dalg24 — APPROVED** [review]

File: containers/src/impl/Kokkos_Functional_impl.hpp (hunk 1 above)
Comment (masterleinad) [reply containers/src/impl/Kokkos_Functional_impl.hpp:93]: We can do
```diff
diff --git a/containers/src/impl/Kokkos_Functional_impl.hpp b/containers/src/impl/Kokkos_Functional_impl.hpp
index 76d57ab2a..43d66a4af 100644
--- a/containers/src/impl/Kokkos_Functional_impl.hpp
+++ b/containers/src/impl/Kokkos_Functional_impl.hpp
@@ -80,7 +80,7 @@ uint32_t MurmurHash3_x86_32(const void* key, int len, uint32_t seed) {
 
   uint32_t k1 = 0;
 
-  switch (len & 3) {
+  switch (len & 3) {  // NOLINT(bugprone-switch-missing-default-case)
     case 3: k1 ^= tail[2] << 16; [[fallthrough]];
     case 2: k1 ^= tail[1] << 8; [[fallthrough]];
     case 1:
@@ -89,8 +89,6 @@ uint32_t MurmurHash3_x86_32(const void* key, int len, uint32_t seed) {
       k1 = rotl32(k1, 15);
       k1 *= c2;
       h1 ^= k1;
-      break;
-    default: break;
   };
 
   //----------
```
instead but I don't have a strong opinion either way (except that the CI is already passing).

Comment (dalg24) [reply containers/src/impl/Kokkos_Functional_impl.hpp:93]: That's what I meant.  Waiting to hear from other reviewers what they think.

Comment (JBludau) [reply containers/src/impl/Kokkos_Functional_impl.hpp:93]: Hmm I think I would go with putting a default like @masterleinad did

Comment (nmm0) [reply containers/src/impl/Kokkos_Functional_impl.hpp:93]: I agree with an empty break; it's not reachable and it's true that the diagnostic is wrong in this case but I'd rather minimize diagnostic suppressions in favor of keeping a default

Comment (masterleinad) [reply containers/src/impl/Kokkos_Functional_impl.hpp:93]: Note that
```C++
[...]
    break;
  default:
};
```
would give me
```
warning: label at end of compound statement is a C++23 extension [-Wc++23-extensions]
```

**Comment by dalg24** [comment]
They are still issues in SIMD that had no coverage in our CI
```
[21/22] Building CXX object simd/unit_tests/CMakeFiles/Kokkos_UnitTest_SIMD.dir/Debug/TestSIMD.cpp.o
FAILED: simd/unit_tests/CMakeFiles/Kokkos_UnitTest_SIMD.dir/Debug/TestSIMD.cpp.o
/Applications/CLion.app/Contents/bin/cmake/mac/aarch64/bin/cmake -E __run_co_compile --tidy="/Users/qdi/.local/clang+llvm-devel/bin/clang-tidy;-warnings-as-errors=*;--extra-arg-before=--driver-mode=g++" --source=/Users/qdi/CLionProjects/kokkos/simd/unit_tests/TestSIMD.cpp -- /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/c++ -DKOKKOS_DEPENDENCE -DCMAKE_INTDIR=\"Debug\" -I/Users/qdi/CLionProjects/kokkos/simd/unit_tests/include -I/Users/qdi/CLionProjects/kokkos/cmake-build-default -I/Users/qdi/CLionProjects/kokkos/cmake-build-default/core/src -I/Users/qdi/CLionProjects/kokkos/core/src -I/Users/qdi/CLionProjects/kokkos/tpls/desul/include -I/Users/qdi/CLionProjects/kokkos/tpls/mdspan/include -I/Users/qdi/CLionProjects/kokkos/cmake-build-default/containers/src -I/Users/qdi/CLionProjects/kokkos/containers/src -I/Users/qdi/CLionProjects/kokkos/cmake-build-default/algorithms/src -I/Users/qdi/CLionProjects/kokkos/algorithms/src -I/Users/qdi/CLionProjects/kokkos/cmake-build-default/simd/src -I/Users/qdi/CLionProjects/kokkos/simd/src -isystem /Users/qdi/.local/googletest-devel/include -Wall -Wextra -Wunused-parameter -Wshadow -pedantic -Wsign-compare -Wtype-limits -Wuninitialized -Wsuggest-override -g -std=c++17 -arch arm64 -isysroot /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX15.2.sdk -fcolor-diagnostics -fno-rtti -mcpu=native -MD -MT simd/unit_tests/CMakeFiles/Kokkos_UnitTest_SIMD.dir/Debug/TestSIMD.cpp.o -MF simd/unit_tests/CMakeFiles/Kokkos_UnitTest_SIMD.dir/Debug/TestSIMD.cpp.o.d -o simd/unit_tests/CMakeFiles/Kokkos_UnitTest_SIMD.dir/Debug/TestSIMD.cpp.o -c /Users/qdi/CLionProjects/kokkos/simd/unit_tests/TestSIMD.cpp
/Users/qdi/CLionProjects/kokkos/simd/src/Kokkos_SIMD_NEON.hpp:457:7: error: switching on non-enum value without default case may not cover all cases [bugprone-switch-missing-default-case,-warnings-as-errors]
  457 |       switch (m_lane) {
      |       ^
/Users/qdi/CLionProjects/kokkos/simd/src/Kokkos_SIMD_NEON.hpp:464:7: error: switching on non-enum value without default case may not cover all cases [bugprone-switch-missing-default-case,-warnings-as-errors]
  464 |       switch (m_lane) {
      |       ^
/Users/qdi/CLionProjects/kokkos/simd/src/Kokkos_SIMD_NEON.hpp:724:7: error: switching on non-enum value without default case may not cover all cases [bugprone-switch-missing-default-case,-warnings-as-errors]
  724 |       switch (m_lane) {
      |       ^
/Users/qdi/CLionProjects/kokkos/simd/src/Kokkos_SIMD_NEON.hpp:731:7: error: switching on non-enum value without default case may not cover all cases [bugprone-switch-missing-default-case,-warnings-as-errors]
  731 |       switch (m_lane) {
      |       ^
/Users/qdi/CLionProjects/kokkos/simd/src/Kokkos_SIMD_NEON.hpp:977:7: error: switching on non-enum value without default case may not cover all cases [bugprone-switch-missing-default-case,-warnings-as-errors]
  977 |       switch (m_lane) {
      |       ^
/Users/qdi/CLionProjects/kokkos/simd/src/Kokkos_SIMD_NEON.hpp:986:7: error: switching on non-enum value without default case may not cover all cases [bugprone-switch-missing-default-case,-warnings-as-errors]
  986 |       switch (m_lane) {
      |       ^
/Users/qdi/CLionProjects/kokkos/simd/src/Kokkos_SIMD_NEON.hpp:1238:7: error: switching on non-enum value without default case may not cover all cases [bugprone-switch-missing-default-case,-warnings-as-errors]
 1238 |       switch (m_lane) {
      |       ^
/Users/qdi/CLionProjects/kokkos/simd/src/Kokkos_SIMD_NEON.hpp:1245:7: error: switching on non-enum value without default case may not cover all cases [bugprone-switch-missing-default-case,-warnings-as-errors]
 1245 |       switch (m_lane) {
      |       ^
/Users/qdi/CLionProjects/kokkos/simd/src/Kokkos_SIMD_NEON.hpp:1453:7: error: switching on non-enum value without default case may not cover all cases [bugprone-switch-missing-default-case,-warnings-as-errors]
 1453 |       switch (m_lane) {
      |       ^
/Users/qdi/CLionProjects/kokkos/simd/src/Kokkos_SIMD_NEON.hpp:1462:7: error: switching on non-enum value without default case may not cover all cases [bugprone-switch-missing-default-case,-warnings-as-errors]
 1462 |       switch (m_lane) {
      |       ^
/Users/qdi/CLionProjects/kokkos/simd/src/Kokkos_SIMD_NEON.hpp:1676:7: error: switching on non-enum value without default case may not cover all cases [bugprone-switch-missing-default-case,-warnings-as-errors]
 1676 |       switch (m_lane) {
      |       ^
/Users/qdi/CLionProjects/kokkos/simd/src/Kokkos_SIMD_NEON.hpp:1683:7: error: switching on non-enum value without default case may not cover all cases [bugprone-switch-missing-default-case,-warnings-as-errors]
 1683 |       switch (m_lane) {
      |       ^
/Users/qdi/CLionProjects/kokkos/simd/src/Kokkos_SIMD_NEON.hpp:1890:7: error: switching on non-enum value without default case may not cover all cases [bugprone-switch-missing-default-case,-warnings-as-errors]
 1890 |       switch (m_lane) {
      |       ^
/Users/qdi/CLionProjects/kokkos/simd/src/Kokkos_SIMD_NEON.hpp:1897:7: error: switching on non-enum value without default case may not cover all cases [bugprone-switch-missing-default-case,-warnings-as-errors]
 1897 |       switch (m_lane) {
      |       ^
3516 warnings generated.
Suppressed 3536 warnings (3502 in non-user code, 34 NOLINT).
Use -header-filter=.* to display errors from all non-system headers. Use -system-headers to display errors from system headers as well.
14 warnings treated as errors
```
`-DKokkos_ARCH_NATIVE=ON` on Apple M1 Pro to reproduce

**Pull request merged**